### ISSUE TITLE
InfluxDB protocol does not properly convert to prometheus metrics (resulting in missing/unselectable data)

### ISSUE THREAD

**Issue by erkexzcx** [VictoriaMetrics/VictoriaMetrics#3113]
I am using Telegraf agent to deliver metrics via InfluxDB protocol to VictoriaMetrics. I am using [below plugin](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/linux_sysctl_fs):
```toml
# Provides Linux sysctl fs metrics
[[inputs.linux_sysctl_fs]]
  # no configuration
```

Here is the example output:
```
> linux_sysctl_fs,host=foo dentry-want-pages=0i,file-max=44222i,aio-max-nr=65536i,inode-preshrink-nr=0i,dentry-nr=64340i,dentry-unused-nr=55274i,file-nr=1568i,aio-nr=0i,inode-nr=35952i,inode-free-nr=12957i,dentry-age-limit=45i 1490982022000000000
```

The issue is here:

- Measurement: `linux_sysctl_fs`
- Field: `file-max`
- Prometheus format: `linux_sysctl_fs_file-max` which is **invalid metric name format**.

Here is the [prometheus metric naming specification](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels). It says:
> It may contain ASCII letters and digits, as well as underscores and colons. It must match the regex `[a-zA-Z_:][a-zA-Z0-9_:]*`.

As you can see [here](https://regex101.com/r/xj250M/1), it does not match given regex and it basically means influxdb metrics ingestion functionality is broken.

**Comment by erkexzcx** [comment]
This can be workarounded in Telegraf. The downside of this is that it will be applied to all other outputs.
```toml
# Workaround for https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3113
[[processors.regex]]
  namepass = ["*"]
  [[processors.regex.field_rename]]
    pattern = "-"
    replacement = "_"
```

**Comment by denisgolius** [comment]
> I am using Telegraf agent to deliver metrics via InfluxDB protocol to VictoriaMetrics. I am using [below plugin](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/linux_sysctl_fs):
> 
> ```toml
> # Provides Linux sysctl fs metrics
> [[inputs.linux_sysctl_fs]]
>   # no configuration
> ```
> 
> Here is the example output:
> 
> ```
> > linux_sysctl_fs,host=foo dentry-want-pages=0i,file-max=44222i,aio-max-nr=65536i,inode-preshrink-nr=0i,dentry-nr=64340i,dentry-unused-nr=55274i,file-nr=1568i,aio-nr=0i,inode-nr=35952i,inode-free-nr=12957i,dentry-age-limit=45i 1490982022000000000
> ```
> 
> The issue is here:
> 
>     * Measurement: `linux_sysctl_fs`
> 
>     * Field: `file-max`
> 
>     * Prometheus format: `linux_sysctl_fs_file-max` which is **invalid metric name format**.
> 
> 
> Here is the [prometheus metric naming specification](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels). It says:
> 
> > It may contain ASCII letters and digits, as well as underscores and colons. It must match the regex `[a-zA-Z_:][a-zA-Z0-9_:]*`.
> 
> As you can see [here](https://regex101.com/r/xj250M/1), it does not match given regex and it basically means influxdb metrics ingestion functionality is broken.

Hi @erkexzcx !
VictoriaMetrics do not convert field names in not Prometheus style because it makes it easier to migrate from InfluxDB queries to Prometheus.

I've added a feature request https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3133 for your case.

**Comment by erkexzcx** [comment]
That's great to hear!

However, you should treat it as a bug and not a feature, because it is **impossible** to use such metrics. It should be mandatory to replace `-` with `_` otherwise those metrics are not selectable in prometheus.

**Comment by denisgolius** [comment]
> That's great to hear!
> 
> However, you should treat it as a bug and not a feature, because it is **impossible** to use such metrics. It should be mandatory to replace `-` with `_` otherwise those metrics are not selectable in prometheus data source in Grafana.

It is not a bug in VictoriaMetrics, it is how PromQL works. To check this you can send a query to VictoriaMetrics via curl and get a proper response, see:

```console
curl -s -G 'http://127.0.0.1:8428/api/v1/export' -d'match={__name__="linux_sysctl_fs_dentry-want-pages"}'
```
Response from your metrics example:
```json
{
  "metric": {
    "__name__": "linux_sysctl_fs_dentry-want-pages",
    "host": "foo"
  },
  "values": [
    0
  ],
  "timestamps": [
    1663654902276
  ]
}
```
As a workaround in Grafana, you can use metrics like `{__name__="linux_sysctl_fs_inode-nr"}` or use escape characters in metrics names `linux_sysctl_fs_dentry\-want\-pages`. See the screenshot:
![image](https://user-images.githubusercontent.com/5650611/191204800-3f7bfd4f-1495-4d27-99ca-f4b0e5c3c16b.png)

**Comment by erkexzcx** [comment]
> As a workaround in Grafana, you can use metrics like {__name__="linux_sysctl_fs_inode-nr"} or use escape characters in metrics names linux_sysctl_fs_dentry\-want\-pages

Excellent! Thank you very much. I did not know about this (tried quotes, but did not work).

Haven't checked yet, but we should also look into https://github.com/influxdata/telegraf/tree/master/plugins/outputs/prometheus_client to see how it gives out influx metrics. I think it would be a great inspiration if it actually converts `-` to `_`.

**Comment by erkexzcx** [comment]
I also think that this should not be ignored:

> It may contain ASCII letters and digits, as well as underscores and colons. It must match the regex [a-zA-Z_:][a-zA-Z0-9_:]*.

At least make that flag enabled by default.

**Comment by erkexzcx** [comment]
Telegraf [Prometheus output plugin ](https://github.com/influxdata/telegraf/tree/master/plugins/outputs/prometheus_client)actually replaces all illegal characters to `_`:
```go
var invalidNameCharRE = regexp.MustCompile(`[^a-zA-Z0-9_:]`)
...
func sanitize(value string) string {
	return invalidNameCharRE.ReplaceAllString(value, "_")
}
```

I expect VictoriaMetrics to do the same approach. Please implement identical approach to VictoriaMetrics.

**Comment by valyala** [comment]
VictoriaMetrics accepts data via [various data ingestion protocols](https://docs.victoriametrics.com/#how-to-import-time-series-data) including [InfluxDB line protocol](https://docs.victoriametrics.com/#how-to-send-data-from-influxdb-compatible-agents-such-as-telegraf). The InfluxDB line protocol [allows ingesting data with almost any chars in metric names, tag names, field names and tag values](https://docs.influxdata.com/influxdb/v1.8/write_protocols/line_protocol_reference/). VictoriaMetrics allows querying time series with any chars in its names, labels and label values - just escape special chars with `\`. That's why by default VictoriaMetrics accepts any chars in metric names, tag names, field values and tag values for data ingested via InfluxDB line protocol.

Sometimes it is needed to replace chars, which aren't allowed by Prometheus in metric names or label names, with underscores like some Prometheus-compatible systems do. That's why VictoriaMetrics may need an additional command-line flag such as `-usePromCompatibleNaming`, which would enable Prometheus-compatible naming for the ingested data. By default this flag must be disabled, since there are many real-world cases, when Prometheus-compatible naming isn't needed for the ingested data.

**Comment by valyala** [comment]
The commit fcffdba9dcbc6a696d328cc8f4268349d54218e3 adds `-usePromCompatibleNaming` command-line flag to `vmagent`, `vminsert` and single-node VictoriaMetrics. This flag can be used for [Prometheus-compatible normalizing](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) of all the metrics and labels ingested via [all the supported data ingestion protocols](https://docs.victoriametrics.com/#how-to-import-time-series-data).

This commit will be included in the next release of VictoriaMetrics.

**Comment by valyala** [comment]
VictoriaMetrics and vmagent support `-usePromCompatibleNaming` command-line flag starting from [v1.82.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the feature request as done.

**Comment by johnseekins** [comment]
@valyala I think you missed an edge case here:

```
Label names may contain ASCII letters, numbers, as well as underscores. They must match the regex [a-zA-Z_][a-zA-Z0-9_]*. Label names beginning with __ are reserved for internal use.
```

Label/tag names can't have `:` in them, but you are allowing them in your regex matching (since you only use the one regex match). The change you made in the opentsdb ingester had the correct regex pattern for tags as well.

**Comment by valyala** [comment]
> Label/tag names can't have : in them, but you are allowing them in your regex matching (since you only use the one regex match). The change you made in the opentsdb ingester had the correct regex pattern for tags as well.

@johnseekins , thanks for the pointer! This has been fixed in the commit fdae53a75b2173c32b673364a4eaf4fb61a21559 . This commit will be included in the next release.

Sorry for late reply - your message has been lost in my email :(

**Comment by valyala** [comment]
The commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/fdae53a75b2173c32b673364a4eaf4fb61a21559 has been included in [v1.93.1](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by valyala** [comment]
FYI, the fix described [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3113#issuecomment-1677825731) has been backported to [v1.87.8 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.8).
